src/main/java/frc/lib/SwerveTrajectoryGenerator.java

```diff
@@ -0,0 +1,279 @@
+                // Turn around in place
+                velocities[i] = 0;
+            } else {
+                velocities[i] = Math.min(
+                        VELOCITY_PROPORTION
+                                / Math.pow(curvatures[i], VELOCITY_POWER),
+                        MAXIMUM_LINEAR_VELOCITY);
+            }
+        }
+        velocities[0] = MAXIMUM_LINEAR_VELOCITY;
+        distances[path.length - 1] = distances[path.length - 2]
+                + path[path.length - 1].minus(path[path.length - 2])
+                                       .getTranslation()
+                                       .getNorm();
+        // Apply deceleration where needed
+        for (int i = path.length - 2; i >= 0; i--) {
+            double newVelocity = Math.sqrt(velocities[i + 1] * velocities[i + 1]
+                    + 2 * MAXIMUM_LINEAR_ACCELERATION
+                            * (distances[i + 1] - distances[i]));
+            if (!Double.isNaN(newVelocity) && newVelocity < velocities[i]) {
+                velocities[i] = newVelocity;
+            }
+        }
+        /*
+         * To project the actual performance of the robot, the commanded
+         * velocities are turned into predicted velocities
+         */
+        System.arraycopy(velocities, 0, predictedVelocities, 0, path.length);
+        predictedVelocities[0] = 0;
+        for (int i = 1; i < path.length - 1; i++) {
+            double newVelocity = Math.sqrt(
+                    predictedVelocities[i - 1] * predictedVelocities[i - 1]
+                            + 2 * MAXIMUM_LINEAR_ACCELERATION
+                                    * (distances[i] - distances[i - 1]));
+            if (!Double.isNaN(newVelocity) && newVelocity < velocities[i]) {
+                predictedVelocities[i] = newVelocity;
+            }
+        }
+    }
+
+    private static Pose2d[] injectAndSmooth(Pose2d[] waypoints) {
+        double[][] points = interpolate(waypoints);
+        double[][] smoothPoints = smooth(points);
+
+        // Convert points back to Pose2d
+        Pose2d[] newPoints = new Pose2d[points.length];
+        for (int i = 0; i < newPoints.length; i++) {
+            newPoints[i] = new Pose2d(smoothPoints[i][0], smoothPoints[i][1],
+                    Rotation2d.fromDegrees(points[i][2]));
+        }
+        return newPoints;
+    }
+
+    private static double[][] interpolate(Pose2d[] waypoints) {
+        // Calculate how many points to insert into each segment and their
+        // initial transforms
+        int[] intermediatePointCounts = new int[waypoints.length - 1];
+        double[][] deltas = new double[waypoints.length - 1][3];
+        int totalPointCount = 1;
+        for (int i = 0; i < deltas.length; i++) {
+            deltas[i][0] = waypoints[i + 1].getX() - waypoints[i].getX();
+            deltas[i][1] = waypoints[i + 1].getY() - waypoints[i].getY();
+            deltas[i][2] = waypoints[i + 1].getRotation()
```

Comment:
In this function you might consider whether assigning some temporaries variables to hold expressions like `waypoints[i+1]` or later `points[pointIndex]` might help readability.